tools/Microsoft.AspNet.Server.Kestrel.GeneratedCode/FrameFeatureCollection.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNet.Http.Features;
+using Microsoft.Dnx.Compilation.CSharp;
+using Microsoft.AspNet.Http.Features.Internal;
+using Microsoft.AspNet.Http.Features.Authentication;
+
+namespace Microsoft.AspNet.Server.Kestrel.GeneratedCode
+{
+    // This project can output the Class library as a NuGet Package.
+    // To enable this option, right-click on the project and select the Properties menu item. In the Build tab select "Produce outputs on build".
+    public class FrameFeatureCollection : ICompileModule
+    {
+        static string Each<T>(IEnumerable<T> values, Func<T, string> formatter)
+        {
+            return values.Select(formatter).Aggregate((a, b) => a + b);
+        }
+
+        public virtual void BeforeCompile(BeforeCompileContext context)
+        {
+            var syntaxTree = Microsoft.CodeAnalysis.CSharp.CSharpSyntaxTree.ParseText(GeneratedFile());
+            context.Compilation = context.Compilation.AddSyntaxTrees(syntaxTree);
+        }
+
+        public static string GeneratedFile()
+        {
+            var commonFeatures = new[]
+            {
+                typeof(IHttpRequestFeature),
+                typeof(IHttpResponseFeature),
+                typeof(IHttpRequestIdentifierFeature),
+                typeof(IHttpSendFileFeature),
+                typeof(IServiceProvidersFeature),
+                typeof(IHttpAuthenticationFeature),
+                typeof(IHttpRequestLifetimeFeature),
+                typeof(IQueryFeature),
+                typeof(IFormFeature),
+                typeof(IResponseCookiesFeature),
+                typeof(IItemsFeature),
+                typeof(IHttpConnectionFeature),
+                typeof(ITlsConnectionFeature),
+                typeof(IHttpUpgradeFeature),
+                typeof(IHttpWebSocketFeature),
+                typeof(ISessionFeature),
+            };
+
+            return $@"
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNet.Server.Kestrel.Http 
+{{
+    public partial class Frame
+    {{
+        {Each(commonFeatures.Select((feature, index) => new { feature, index }), entry => $@"
+        private const long flag{entry.feature.Name} = {1 << entry.index};")}
+
+        {Each(commonFeatures, feature => $@"
+        private static readonly Type {feature.Name}Type = typeof(global::{feature.FullName});")}
+
+        private long _featureOverridenFlags = 0L;
```

Comment:
Updated, am using ```_featureOverridenFlags``` for ```IHttpRequestFeature```, ```IHttpResponseFeature```, ```IHttpUpgradeFeature``` which are generally ```this``` so use one int rather than 3 pointers for it.